src/hotspot/share/gc/g1/g1ServiceThread.hpp

```diff
@@ -28,43 +28,91 @@
+  G1ServiceTask* _next;
+  // The service thread this task is associated with.
+  G1ServiceThread* _service_thread;
+
+public:
+  G1ServiceTask(const char* name);
+  const char* name();
+  void set_service_thread(G1ServiceThread* thread);
+
+  void set_time(jlong time);
+  jlong time();
+
+  void set_next(G1ServiceTask* next);
+  G1ServiceTask* next();
+
+  // Do the actual work for the task. To get added back to the
+  // execution queue a task can call schedule(delay_ms).
+  virtual void execute() = 0;
+
+protected:
+  // Schedule the task on the associated service thread
+  // using the provided delay in milliseconds.
+  void schedule(jlong delay_ms);
+};
+
+class G1SentinelTask : public G1ServiceTask {
+public:
+  G1SentinelTask();
+  virtual void execute();
+};
+
+class G1ServiceTaskQueue {
+  // The sentinel task is the entry point of this priority queue holding the
+  // service tasks. The queue is ordered by the time the tasks are scheduled
+  // to run and the sentinel task has the time set to DBL_MAX. This guarantees
+  // that any new task will be added just before the sentinel at the latest.
+  G1SentinelTask _sentinel;
+
+  // Verify that the queue is ordered.
+  void verify_task_queue() NOT_DEBUG_RETURN;
+public:
+  G1ServiceTaskQueue();
+  G1ServiceTask* pop();
+  G1ServiceTask* peek();
+  void add_ordered(G1ServiceTask* task);
+  bool is_empty();
+};
+
 // The G1ServiceThread is used to periodically do a number of different tasks:
 //   - re-assess the validity of the prediction for the
 //     remembered set lengths of the young generation.
 //   - check if a periodic GC should be scheduled.
 class G1ServiceThread: public ConcurrentGCThread {
-private:
+  // The monitor is used to ensure thread safety for the task queue
+  // and allow other threads to signal the service thread to wake up.
   Monitor _monitor;
-
-  double _last_periodic_gc_attempt_s;
+  G1ServiceTaskQueue _task_queue;
 
   double _vtime_accum;  // Accumulated virtual time.
 
-  // Sample the current length of remembered sets for young.
-  //
-  // At the end of the GC G1 determines the length of the young gen based on
-  // how much time the next GC can take, and when the next GC may occur
-  // according to the MMU.
-  //
-  // The assumption is that a significant part of the GC is spent on scanning
-  // the remembered sets (and many other components), so this thread constantly
-  // reevaluates the prediction for the remembered set scanning costs, and potentially
-  // G1Policy resizes the young gen. This may do a premature GC or even
-  // increase the young gen size to keep pause time length goal.
-  void sample_young_list_rs_length();
-
   void run_service();
-  void check_for_periodic_gc();
-
   void stop_service();
 
+  // Returns the time in milliseconds until the next task is due.
+  // Used both to determine if there are tasks ready to run and
+  // how long to sleep when nothing is ready.
+  int64_t time_to_next_task_ms();
```

Comment:
I recommend returning a positive number here because it only raises questions what this means. The implementation of the method also never returns negative values.